### DISCUSSION TITLE
Allow integrations to update capabilities of entities

### DISCUSSION THREAD

**Discussion by emontnemery** [home-assistant/architecture#985] [Entity Models] [answered]
## Proposal

Allow integrations to update capabilities of entities: `capability_attributes`, `device_class`, `supported_features`

## Rationale

Updating capabilities of entities is in some cases a valid use case:
- We allow changing device class from the UI for some entities
- Reconfiguring a device may cause it's capabilities to change
- Capabilities for a device may change depending on it's mode, as an example the supported temperature interval of a `climate` entity may be different depending on its mode

## Background

Architecture issue https://github.com/home-assistant/architecture/issues/1 concluded that it's not allowed for an integration to change the `supported_features` of an entity, the main concern seems to be that it breaks integration with Alexa. Although `capability_attributes` and `device_class` are not mentioned in #1, they too are crucial for integration with Alexa and other voice assistants. However, our voice assistant integrations have become much smarter since February 2018 when that issue was opened; Alexa, Assist and Google Assistant handle changes in entity capabilities and there's an open PR for Home Kit to do the same: https://github.com/home-assistant/core/pull/101719

## Concerns

Issue #1 in particular focuses on media players dynamically changing their supported features according to the media currently playing etc. That's perhaps still not a wanted behavior, i.e. a media player should set its supported features to the union of all possible supported features? If we allow integrations to update capabilities, should we give some guidelines to where capabilities should and should not be updated?

### Mitigation

We should add rate limit to prevent integrations from thrashing the entity registry. Details will be determined in the implementation PR.

**Comment by thecode** [comment]
I think it is good to allow integrations to update the capabilities, since 2018 many things have evolved.
For climate this is very important, many climate devices support different options according to the HVAC mode.

**Comment by jbouwh** [comment]
The features of MQTT entities when discovered can be updated automatically already. It would be good this would be supported and optimized.

**Comment by emontnemery** [comment, marked as answer]
The proposal is implemented in core PR https://github.com/home-assistant/core/pull/101748
